### PR TITLE
feat: add multifernet

### PR DESCRIPTION
**Pull request by bbangert**
Committed on behalf of anonymous author.

### CODE REVIEW

**Comment by alex** [comment]
Could the cargo-fmt changes be factored out into their own PR? They make this more difficult to review.

File: src/lib.rs [hunk 1]
```diff
@@ -23,18 +22,48 @@ pub struct Fernet {
 #[derive(Debug, PartialEq, Eq)]
 pub struct DecryptionError;
 
+pub struct MultiFernet {
+    fernets: Vec<Fernet>,
+}
+
+impl MultiFernet {
+    pub fn new(keys: Vec<String>) -> MultiFernet {
```
Comment (alex) [inline src/lib.rs]: This should just take a `Vec<Fernet>` like we do in the Python version.

There should also be an `assert`/`panic` that there's at least one element in the vec

File: src/lib.rs [hunk 2]
```diff
@@ -220,13 +291,17 @@ mod tests {
...
     }
 
+    #[test]
+    fn test_multi_roundtrips() {
+        let f = MultiFernet::new(vec![base64::encode_config(&vec![0; 32], base64::URL_SAFE)]);
+
+        for val in [b"".to_vec(), b"Abc".to_vec(), b"\x00\xFF\x00\x00".to_vec()].into_iter() {
+            assert_eq!(f.decrypt(&f.encrypt(&val)), Ok(val.clone()));
+        }
+    }
```
Comment (alex) [inline src/lib.rs]: More tests needed: https://github.com/pyca/cryptography/blob/master/tests/test_fernet.py#L134 contains examples of the types of tests I'd like to see

Comment (bbangert) [reply src/lib.rs]: Added the other tests. Didn't add the rotate tests as the TTL stuff isn't there yet.

File: src/lib.rs [hunk 3]
```diff
@@ -39,6 +40,34 @@ pub struct Fernet {
...
+impl MultiFernet {
+    pub fn new(keys: Vec<Fernet>) -> MultiFernet {
+        assert!(keys.len() > 0);
+        MultiFernet {
+            fernets: keys,
+        }
+    }
+
+    pub fn encrypt(&self, data: &[u8]) -> String {
+        self.fernets.first().map(|f| f.encrypt(data)).unwrap()
```
Comment (alex) [inline src/lib.rs]: Can this simplify to `self.fernets[0].encrypt(data)`?

File: src/lib.rs [hunk 4]
```diff
@@ -336,4 +365,37 @@ mod tests {
             assert_eq!(f2.decrypt(&f1.encrypt(&val)), Ok(val.clone()));
         }
     }
+
+    #[test]
+    fn test_multi_encrypt() {
+        let f1 = Fernet::new(&Fernet::generate_key()).unwrap();
+        let f2 = Fernet::new(&Fernet::generate_key()).unwrap();
+        let f = MultiFernet::new(vec![f1, f2]);
+        assert_eq!(f1.decrypt(&f.encrypt(&b"abc".to_vec())).unwrap(), b"abc".to_vec());
```
Comment (alex) [inline src/lib.rs]: You can drop the `&` + `to_vec()` on the inner `b"abc"`.

File: src/lib.rs (hunk 3 above)
Comment (bbangert) [reply src/lib.rs]: Yup, fixed.

File: src/lib.rs (hunk 4 above)
Comment (bbangert) [reply src/lib.rs]: Ah right, cool.

File: src/lib.rs [hunk 5]
```diff
@@ -336,4 +372,37 @@ mod tests {
             assert_eq!(f2.decrypt(&f1.encrypt(&val)), Ok(val.clone()));
         }
     }
+
+    #[test]
+    fn test_multi_encrypt() {
+        let f1 = Fernet::new(&Fernet::generate_key()).unwrap();
+        let f2 = Fernet::new(&Fernet::generate_key()).unwrap();
+        let f = MultiFernet::new(vec![f1, f2]);
+        assert_eq!(f1.decrypt(&f.encrypt(b"abc")).unwrap(), b"abc".to_vec());
```
Comment (alex) [inline src/lib.rs]: Can you add an `assert_eq!(f2.decrypt(&f.decrypt(b"abc")), Err(DecryptionError));`?

last comment, I swear :-)

Comment (bbangert) [reply src/lib.rs]: No, but I can add ```assert_eq!(f2.decrypt(&f.encrypt(b"abc")), Err(DecryptionError));```. ;)

Comment (alex) [reply src/lib.rs]: :D

File: src/lib.rs [hunk 6]
```diff
@@ -29,6 +29,7 @@ use rand::Rng;
 
 const MAX_CLOCK_SKEW: u64 = 60;
 
+#[derive(Clone, Copy)]
```
Comment (alex) [inline src/lib.rs]: Err, wait, why are these required? I think `Fernet` should probably be move-only.

Comment (bbangert) [reply src/lib.rs]: Cause I didn't want to copy the key and make more identical Fernets myself, ie, the tests.

Comment (alex) [reply src/lib.rs]: It doesn't seem to be actually used though?

If possible I'd prefer not to have this, it simplifies reasoning about where in memory your keys are.

Comment (bbangert) [reply src/lib.rs]: Heh, well, until you through them into a MultiFernet and they jump onto the heap in the Vec. But sure, I can remove that, it just makes the tests a bit more annoying.

**Comment by alex** [comment]
I think it'd be good to drop the Copy/Clone. It means if we ever add a
secure-erase drop implementation it'll be easier.

**Comment by bbangert** [comment]
Dropped the Copy/Clone.
